packages/react-native/React/Views/RCTViewManager.m

```diff
@@ -175,12 +176,24 @@ - (RCTShadowView *)shadowView
 
 RCT_CUSTOM_VIEW_PROPERTY(transform, CATransform3D, RCTView)
 {
-  view.layer.transform = json ? [RCTConvert CATransform3D:json] : defaultView.layer.transform;
+  CATransform3D _transform = json ? [RCTConvert CATransform3D:json] : defaultView.layer.transform;
+  if (view.transformOrigin != nil) {
+      applyTransformWithTransformOrigin(view, view.transformOrigin, _transform);
+  } else {
+      view.layer.transform = _transform;
+  }
+  
   // Enable edge antialiasing in rotation, skew, or perspective transforms
   view.layer.allowsEdgeAntialiasing =
       view.layer.transform.m12 != 0.0f || view.layer.transform.m21 != 0.0f || view.layer.transform.m34 != 0.0f;
 }
 
+RCT_CUSTOM_VIEW_PROPERTY(transformOrigin, NSString*, RCTView)
+{
+    view.transformOrigin = json;
+    applyTransformWithTransformOrigin(view, json, view.layer.transform);
```

Comment:
The order of setters being called is not defined in RN. How are you making sure you're not changing the transform origin of a transform that already had it changed?

Reply:
I had this concern and i think we can make it better. Right now to make it order-independent I am setting `view.transformOrigin` instance variable on the view. If the transform setter gets called first it'll perform the transform (because it doesn't know there will be a transformOrigin call too!), later transform origin setter will reuse the same layer transform and perform it again. Subsequent calls to transform setter will use the `view.transformOrigin` variable and perform transforms. Subsequent calls to transformOrigin setter will reuse existing layer transform and re-apply the transforms and also set the `view.transformOrigin` variable.

The problem with this approach is that transform gets applied twice during initial mount and we also need an instance variable. If we could use a single setter e.g. `transform: { transformOrigin: "top", transforms: [{rotate: "45deg"}]}` we can optimise it.

Reply:
In the old architecture, you could use `didSetProps` to run some additional transactions after all props have been set.